### ISSUE TITLE
Symmetric data become asymmetric in a heatmap

### ISSUE THREAD

**Issue by piccolbo** [vega/vega-lite#4133]
Please:
- [ x] Check for duplicate issues. Please file separate requests as separate issues on GitHub.
- [ x] Describe how to reproduce the bug / the goal of the feature request.
- [ x] Use the latest versions of Vega and Vega-Lite.
- [ x] Provide an example spec in JSON, wrapped by triple backticks like this:
hope a [link](https://vega.github.io/editor/#/url/vega-lite/N4KABGBEAkDODGALApgWwIaQFxUQFzwAdYsB6UgN2QHN0A6agSz0QFcAjOxge1IRQyUa6ALQAbZskoAmOgBY6AZjoArWNwB2kADTgo8TQDNG1bGFAQIkCo2QB3MxctWUJ-GcUAGT7udW7jAAmLGZy3noQAL56kb5Qgeh4mDhOVhToYqzIsGYA2hGWqX6QAB5mIgCscX5QAJ7lVQXOkABeZhWeTWCxXUXNZTgictXF9YONNVZtOHIAjF09NX2WpeWKI81jYJUbK9NginILu8tWA9vSu1ZbO11TZtIAnMe9d1DnIrNXdQ3frQ8ANheSzeqxwPlBNwmk3+OGkFWBflO7zMX0hv1B+2kQJqiyRoPOl3R4z+WOeuJOBI8fyhpI8Rwpr0mKJmNIxzNhYDmiOcyLBYGhNUgtMx7U6jJBHI+gtG5WGopm8wl+KlcrZg3lHP2imkPMKVMG62J201MKxCOVvINF3VJrpcPFfjxVtVgzRHJupqF+1mAHY9RA+ecIR61QqwLMcU7Ka6I7ahvaI-7LfrY0TQxrE9JHc5namYecjRm7eH4QHzNavbLM+GdeWg+141XmvtuSnA9bbsWREWzfT653mysbr3vQ8LdGmQW1vHR8UfQAOAexkTpmEjxOzRTL6du2ebnOWPMd2Mh9cz8PkyeS3dx409g87oWE-fhrdP4qF19a1FL9sV2Mh2uC8fzhCdcxjW8ZU2EC+xwQ4P36dlz0GNcxzhK8IKnZ8w27ND5weQ8okgnDDXjfCW1Rbd-wbVDyMTP9rxVW9Pno8NwKPEjPzMM8hRuCi9lCRCVnOd0UJtdjhLOB42NAsBGKwm9SIOWS4IjaimJdW8gJ+OjSyI7ouKQnBoOHcoBPuDCpJZbZTOAvdSyjRTmOUhN7zE9Ck2s-kH3cx8aM7CzdO2DyCJM7yPlCmCHLk3UAtPeMosEnAlU0-NlKS+yQqzCKZL8yT4tvOdouyt8DOPADtMSzdkzSk8oOqxzcpJY1eLCsAy0K1ydOFHjNyczjsO4sjWsTTDBqU4aJOLNrKJmZrSpmzcFtmszwUTcqjJE1FbVWyyIwWoLevW8MGTqyrlOKtawD2qB9nG4ihuMrldv6ha7OC27OU687aNs21MruwEFrc4tAc5P0QaurLwZ9Tanu2vSwcTDjHsm57WONWGHhWgHlq6qbwa2bG4UOvGCt+61oeCkmI3h9HEZerGave8nYoGtGXKmrtxKOskQZ6rY+cIqHbWFlKNOcrTXKOoWGJBom8rk1HDIR6STuLcWuQWxW4RRsnjS1hSJq557qeOjrN0lk3pamwWlbU7NWcNrMHtVhn1f+41ze1M6pfS7n7fgrMVYqv7fOLH3fxB2XqTfa3OdtjHdZUsrce9saddtKOUvp03Gdj4P44W82thziNjcTgPnqD1PYtDrbPY+i3y4QgmMebrYetbVL-fq7rbW7jw4spldS9CEOY8HrM86TxnMeLIeUtqvuLqmr6u7e9vGZTpevO3z3C6Z2LZ+rxnx9ZUsG7Vmza73usD5szuJ9O3ubbPz2eb4xt2NP-vA9tM3HuotjRAP7I-HyR8wFWQgZFQBWYObu3zp7DeP965Z1AQgg2xZoEdTdmHKm8Dax+3fv-GuRC5JtlHg1TBv8FgFAALoxDiJAZAGgDCBEYBoUwKQCiQAMGIbgAAnRwXRIDGGQGIQIZh-i7EgAgDIyBREcn4GgJROBIDUCEcgNh7BMhKPLJAPAtRCDqKgAAR1YOgDQeBmCJEYFQSABQ8z8mWOI2wUiZFlDkcY0xMjLHWNsUkWxjjnEjAtm4iRniNH1B8SYsxkAAk2LsSEgxuZmF6EgBgIRABrGR2j4B4CcZEIAA) will do

- [x] If applicable, include error messages and screenshots, GIF videos (e.g. using https://www.cockos.com/licecap/), or working examples (e.g. https://bit.ly/vega-lite-blocks)

Vega editor should show the picture. The data is symmetrical around vertical and horizontal axis, centered in 0,0. If the field type of x and y fields is ordinal, then the problem doesn't manifest itself (but the data is quantitative, not ordinal, and with ordinal  one gets a grid of squares, not a heatmap, additionally there are examples in the gallery with quantitative fields x and y that work just fine). With quantitative for x and y, only the bottom left quadrant displays correctly, the top left and bottom right ones look like the data is constant along one dimension. The top right quadrant look constant, clamped to the value of the very top right corner. I am new to vega-lite and vega so apologies if this is a misunderstanding on my part.

Thanks!

**Comment by piccolbo** [comment]
Thanks Dominik, but I still don't understand.  What is setting the size of mark_rect? I have only two channels and I need four to set two opposite corners. Plus here is an example with quantitative x and y but no overlap:

https://vega.github.io/vega-lite/examples/rect_binned_heatmap.html

But they are binned. Maybe that's the key to what  I am missing.  Let's say my data is lat, long, pressure. I have to make lat and long ordinal for it to work? That's a hack. Lat is a real number. Thanks.

**Comment by domoritz** [comment]
Yes, the data is binned and so Vega-Lite can automatically set the size correctly. 

If you want to use quantitative, you can do that. Instead of just setting x and y, you also need to set x2 and y2. Does that make sense?

**Comment by piccolbo** [comment]
Yes. Somehow the documentation failed me, or I failed on my own.

**Comment by domoritz** [comment]
How can we improve the documentation? If you have a suggestion, please send us a pull request.

**Comment by piccolbo** [comment]
I am pretty new to this stuff so I am weary of telling people how to do things. On the other hand in R I am a ggplot wizard or close, so it's not like this grammar thing is new to me. I am building on top of altair, so I look at the docs there, then if I am desperate I go to vega-lite.  It may be relevant or not to vega-lite but this is my failure path. Altair doc for mark_rect are nada de nada. They send you to MarkDef, which mixes all the marks. Cognitive load here approaches +inf. MarkDef has no mention of which channels are relevant. 0 zip nada. So I go to the vega-lite docs hoping to reverse map to altair. And now this is relevant to this project. This is what i found

https://vega.github.io/vega-lite/docs/rect.html#rect-mark-properties


And specifically:

```
{
  ...
  "mark": {
    "type": "rect",
    ...
  },
  "encoding": ... ,
  ...
}
```
Am I supposed to get anything out of that? Yes if you read past the examples it does mention x2, y2 talking about something, my bad for missing it. But is there anywhere a matrix channel x mark that shows what affects what, and then a link to how the channel affects the mark? If there is, I haven' found it yet.

**Comment by domoritz** [comment]
Thank you for taking the time to explain your path. This is super helpful to improve the documentation. 

I guess the altair docs could have an example of ranged rectangles similar to https://vega.github.io/vega-lite/docs/rect.html#ranged. The Vega-Lite docs could mention more prominently that if you use rect with quantitative, you will get overlapping marks. Would that help?

Please understand that we are volunteers and writing docs is really really hard.

**Comment by piccolbo** [comment]
Totally understand.  Examples never hurt but it seems to me that here we need some invariant in the docs. Encodings do have a list of marks that are affected by them (https://vega.github.io/vega-lite/docs/encoding.html#position) but marks don't have a list of relevant encodings -- or I haven't found it. And there has to be a place to explain the "interaction"  between a mark and an encoding: what does x2 mean to a rect vs and area? Maybe it's a naming issue. Maybe a rect should have a left and right, not an x and x2. Then it would explain itself.

**Comment by jakevdp** [comment]
Good points @piccolbo – I consider https://altair-viz.github.io/user_guide/marks.html a bit of a stub. It would be useful to have more discussion of considerations when using various marks on that page. If that's something you'd be willing to contribute to the documentation, we would greatly appreciate it!
